lms/db/_bulk_action.py

```diff
@@ -84,14 +89,32 @@ def upsert(self, model_class, values):
 
         config = model_class.BULK_CONFIG
 
+        upsert_update_elements = list(config.upsert_update_elements)
+
+        if config.upsert_trigger_onupdate:
+            onupdate_columns = self._get_columns_onupdate(model_class)
+            for column_name, onupdate_value in onupdate_columns:
+                upsert_update_elements.append(column_name)
+                # Copy the values, we don't want to mess with the caller's data
+                values = deepcopy(values)
+                for value_ in values:
+                    value_[column_name] = (
+                        # SQL alchemy wraps functions passed to onupdate or default and could potentially take a "context" argument
+                        # getting a suitable context at this point of the execution it's not possible so we don't support it so we just pass None
+                        # https://docs.sqlalchemy.org/en/14/core/defaults.html#context-sensitive-default-functions
+                        onupdate_value(None)
+                        if callable(onupdate_value)
+                        else onupdate_value
+                    )
```

Comment:
```suggestion
                # SQL alchemy wraps functions passed to onupdate or default and
                # could potentially take a "context" argument getting a
                # suitable context at this point of the execution it's not
                # possible so we don't support it so we just pass None
                # https://docs.sqlalchemy.org/en/14/core/defaults.html#context-sensitive-default-functions
                default_value = (
                    onupdate_value(None) if callable(onupdate_value) else onupdate_value
                )

                # Copy the values, we don't want to mess with the caller's data
                values = deepcopy(values)
                for row in values:
                    row[column_name] = default_value
```

I think the default value is the same for every row, so we don't need to calculate it once per row.

Also, I'm not sure why `value_` has an underscore on the end. Is this because `values` is a list of dicts and it's confusing? I used `row` in the example above which I think makes this clearer?